// oneentry/dist/base/stateModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class StateModule {
    constructor(url, config) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        this.url = url;
        this.lang = (_a = config.langCode) !== null && _a !== void 0 ? _a : 'en_US';
        this.token = config.token;
        this.multipleResponse = !!((_b = config.traficLimit) !== null && _b !== void 0 ? _b : true);
        this.refreshToken = (_d = (_c = config.auth) === null || _c === void 0 ? void 0 : _c.refreshToken) !== null && _d !== void 0 ? _d : undefined;
        this.customAuth = (_f = (_e = config.auth) === null || _e === void 0 ? void 0 : _e.customAuth) !== null && _f !== void 0 ? _f : false;
        this.saveFunction = (_h = (_g = config.auth) === null || _g === void 0 ? void 0 : _g.saveFunction) !== null && _h !== void 0 ? _h : null;
        try {
            if (typeof process === 'object' && +process.versions.node.split('.')[0] < 18) {
                try {
                    this._NO_FETCH = true;
                    this._https = require('https'); //file extension
                }
                catch (e) {
                    this._https = null;
                }
            }
            else {
                this._NO_FETCH = false;
                this._https = null;
            }
        }
        catch (e) {
            this._NO_FETCH = false;
            this._https = null;
        }
    }
}
exports.default = StateModule;
